pkg/trie/triedb/proof/generate.go

```diff
@@ -0,0 +1,455 @@
+					keyNibbles,
+					len(entry.prefix),
+					recordedNodes,
+				)
+
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			switch s := nextStep.(type) {
+			case stepDescend:
+				childPrefix := keyNibbles[:s.childPrefixLen]
+				var childEntry *stackEntry
+				switch child := s.child.(type) {
+				case nodeHandleHash:
+					childRecord := recordedNodes.Next()
+
+					if !bytes.Equal(childRecord.Hash, child.hash) {
+						panic("hash mismatch")
+					}
+
+					outputIndex := len(proofNodes)
+
+					// Insert a placeholder into output which will be replaced when this
+					// new entry is popped from the stack.
+					proofNodes = append(proofNodes, []byte{})
+					childEntry, err = newStackEntry(
+						childPrefix,
+						childRecord.Data,
+						&childRecord.Hash,
+						&outputIndex,
+					)
+
+					if err != nil {
+						return nil, err
+					}
+				case nodeHandleInline:
+					if len(child.data) > common.HashLength {
+						return nil, errors.New("invalid hash length")
+					}
+					childEntry, err = newStackEntry(
+						childPrefix,
+						child.data,
+						nil,
+						nil,
+					)
+					if err != nil {
+						return nil, err
+					}
+				}
+				stack.PushBack(childEntry)
+			default:
+				recordedNodes.Next()
+				break loop
+			}
+		}
+	}
+
+	err = unwindStack(stack, proofNodes, nil)
+	if err != nil {
+		return nil, err
+	}
+	return proofNodes, nil
+}
+
+// / Unwind the stack until the given key is prefixed by the entry at the top of the stack. If the
+// / key is nil, unwind the stack completely. As entries are popped from the stack, they are
+// / encoded into proof nodes and added to the finalized proof.
+func unwindStack(
+	stack *deque.Deque[*stackEntry],
+	proofNodes [][]byte,
+	maybeKey *[]byte,
+) error {
+	for stack.Len() > 0 {
+		entry := stack.PopBack()
+		if maybeKey != nil && bytes.HasPrefix(*maybeKey, entry.prefix) {
+			stack.PushBack(entry)
+			break
+		}
+
+		if stack.Len() > 0 {
+			parentEntry := stack.Back()
+			if parentEntry != nil {
+				encoded, err := entry.encodeNode()
+				if err != nil {
+					return err
+				}
+				parentEntry.setChild(encoded)
+			}
+		}
+
+		index := entry.outputIndex
+		if index != nil {
+			encoded, err := entry.encodeNode()
+			if err != nil {
+				return err
+			}
+			proofNodes[*index] = encoded
+		}
+	}
+	return nil
+}
+
+func sortAndDeduplicateKeys(keys []string) []string {
+	slices.Sort(keys)
+
+	if len(keys) == 0 {
+		return keys
+	}
+
+	result := []string{keys[0]}
+	for i := 1; i < len(keys); i++ {
+		if keys[i] != keys[i-1] {
+			result = append(result, keys[i])
+		}
+	}
+
+	return result
+}
+
+func matchKeyToNode(
+	node codec.EncodedNode,
+	omitValue *bool,
+	childIndex *int,
+	key []byte,
+	prefixlen int,
+	recordedNodes *triedb.RecordedNodesIterator,
+) (step, error) {
+	switch n := node.(type) {
+	case codec.Empty:
+		return stepFoundValue{nil}, nil
+	case codec.Leaf:
+		if bytes.Contains(key, n.PartialKey) && len(key) == prefixlen+len(n.PartialKey) {
```

Comment:
can we use `hasPrefix`? I wonder if exists a possibility of `key` contains the `n.PartialKey` in its middle and end up passing this check

Reply:
We can't, since `n.partialKey` is not the prefix.
For example:
Key could be `6,1,6,2`
and this leaf PartialKey could be `2`

It is most like a postFix, the second part of the condition is to check what you say to prevent this being in the middle